File: 198-house-robber/house-robber.cpp
class Solution {
public:
int find(int i,vector<int>&v,vector<int>&nums){
    if(i==0) return nums[i];
    if(i<0) return 0;
    if(v[i]!=-1) return v[i];
    int take =find(i-2,v,nums)+nums[i];
    int nottake=find(i-1,v,nums);
    
     return v[i]=max(take,nottake);
}
    int rob(vector<int>& nums) {
        
        int i=nums.size()-1;
        vector<int>v(i+1,-1);
        return find(i,v,nums);
    }
};

File: 198-house-robber/house-robber.java
class Solution {
     int find(int i,int[]dp,int[] h){
        if(i==0) return h[0];
        if(i==1) return Math.max(h[0],h[1]);
        if(dp[i]!=-1)return dp[i];
        int take=find(i-2,dp,h)+h[i];
        int nottake=find(i-1,dp,h);
        return dp[i]=Math.max(take,nottake);
    }
    public int rob(int[] h) {
         int n=h.length;
        int[]dp=new int[n+1];
        Arrays.fill(dp,-1);
        return find(n-1,dp,h);
    }
}